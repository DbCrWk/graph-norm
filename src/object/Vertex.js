// @flow
import doesSet from '../func/doesSet';

export type Label = string;

class Vertex {
    label: Label;

    inbound: Set<Label>;

    outbound: Set<Label>;

    constructor(label: string | number) {
        this.label = label.toString();
        this.inbound = new Set();
        this.outbound = new Set();
    }

    isSameAs(y: Vertex): boolean {
        if (this.label !== y.label) return false;
        if (!doesSet(this.inbound).equal(y.inbound)) return false;
        if (!doesSet(this.outbound).equal(y.outbound)) return false;

        return true;
    }
}

export default Vertex;
