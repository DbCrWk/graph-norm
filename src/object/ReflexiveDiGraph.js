// @flow
import Vertex from './Vertex';
import type { Label } from './Vertex';

class ReflexiveDiGraph {
    // TODO: this type could be made stronger to enforce that a NodeLabel always
    // maps to the right NodeLabel
    vertices: { [Label]: Vertex };

    constructor() {
        this.vertices = {};
    }

    addVertex(vertex: Vertex) {
        const { label } = vertex;

        const newVertex = new Vertex(label);
        newVertex.outbound.add(label);
        newVertex.inbound.add(label);

        this.vertices[label] = newVertex;
    }

    addEdge(vertexA: Vertex, vertexB: Vertex) {
        if (!(vertexA.label in this.vertices)) this.addVertex(vertexA);
        if (!(vertexB.label in this.vertices)) this.addVertex(vertexB);

        this.vertices[vertexA.label].outbound.add(vertexB.label);
        this.vertices[vertexB.label].inbound.add(vertexA.label);
    }

    isSameAs(h: ReflexiveDiGraph): boolean {
        const gVertexLabels = Object.keys(this.vertices);
        const hVertexLabels = Object.keys(h.vertices);

        if (gVertexLabels.length !== hVertexLabels.length) return false;
        const isGVertexSameAsH = gVertexLabels.map(
            u => (u in h.vertices) && (h.vertices[u].isSameAs(this.vertices[u])),
        );
        const areAllVerticesTheSame = isGVertexSameAsH.reduce((p, c) => p && c);

        return areAllVerticesTheSame;
    }
}

export default ReflexiveDiGraph;
