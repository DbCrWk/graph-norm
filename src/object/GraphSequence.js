// @flow
import ReflexiveDiGraph from './ReflexiveDiGraph';
import x from '../func/x';
import { debugLib as debugGn } from '../util/logger';

const namespace = 'Object > GraphSequence';
const debug = debugGn(namespace);

class GraphSequence {
    sequence: Array<ReflexiveDiGraph>;

    cumulantSequence: Array<ReflexiveDiGraph>;

    length: number;

    constructor() {
        this.sequence = [];
        this.cumulantSequence = [];
        this.length = 0;
    }

    push(g: ReflexiveDiGraph) {
        debug('.push', 'Starting', { label: g.label });
        this.sequence.push(g);
        this.length += 1;
        debug('.push', 'Graph inserted into sequence', { length: this.length });

        if (this.sequence.length === 1) {
            debug('.push', 'First graph detected; setting as start of cumulantSequence');
            this.cumulantSequence.push(g);
            return;
        }

        const previousGraph = this.getLastCumulant();
        debug('.push', 'Previous graph found', { label: g.label, previousGraph: { label: previousGraph.label } });

        debug('.push', 'Generating product graph');
        const productGraph = x(previousGraph, g);
        debug('.push', 'Product graph created', { label: g.label, productGraph: { label: productGraph.label } });


        debug('.push', 'Pushing to cumulant sequence');
        this.cumulantSequence.push(productGraph);
        debug('.push', 'Done', { label: g.label });
    }

    split(p: number): Array<GraphSequence> {
        debug('.split', 'Starting');

        const head = new GraphSequence();
        const point = new GraphSequence();
        const tail = new GraphSequence();

        this.sequence.forEach((g, i) => {
            if (i < p) { head.push(g); return; }
            if (i === p) { point.push(g); return; }
            tail.push(g);
        });

        debug('.split', 'Done');
        return [head, point, tail];
    }

    isTransitive(): boolean {
        const lastCumulant = this.getLastCumulant();
        return lastCumulant.isTransitive();
    }

    getLastCumulant(): ReflexiveDiGraph {
        return this.cumulantSequence[this.cumulantSequence.length - 1];
    }
}

export default GraphSequence;
