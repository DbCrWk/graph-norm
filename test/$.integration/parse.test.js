// @flow
import ReflexiveDiGraph from '../../src/object/ReflexiveDiGraph';
import GraphSequence from '../../src/object/GraphSequence';
import SequenceTemporalNode from '../../src/object/Sequence.TemporalNode';
import Vertex from '../../src/object/Vertex';
import RootTopologyNode from '../../src/object/Root.TopologyNode';

describe('parsing', () => {
    it('correctly produces the temporal and topological trees', () => {
        expect.assertions(4);

        const a = new Vertex('a');
        const b = new Vertex('b');
        const c = new Vertex('c');

        const g = new ReflexiveDiGraph('g');
        g.addEdge(a, b);
        g.addEdge(b, c);

        const h = new ReflexiveDiGraph('h');
        h.addVertex(a);
        h.addEdge(b, c);
        h.addEdge(c, b);

        const seq = new GraphSequence();
        seq.push(g);
        seq.push(h);
        seq.push(h);
        seq.push(h);

        const temporalTree = new SequenceTemporalNode(seq);
        const topologicalTree = new RootTopologyNode(temporalTree);

        expect(temporalTree).not.toBeNull();
        expect(topologicalTree).not.toBeNull();

        expect(temporalTree.toNode()).toMatchSnapshot();
        expect(topologicalTree.toNode()).toMatchSnapshot();
    });
});
