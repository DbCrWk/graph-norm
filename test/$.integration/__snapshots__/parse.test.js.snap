// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parsing correctly produces the temporal and topological trees 1`] = `
Object {
  "category": "sequence",
  "children": Array [
    Object {
      "category": "sequence",
      "children": Array [
        Object {
          "category": "graph",
          "isTransitive": false,
          "label": "g",
        },
      ],
      "isTransitive": false,
      "label": "(g ... g)",
    },
    Object {
      "category": "graph",
      "isTransitive": true,
      "label": "h",
    },
    Object {
      "category": "sequence",
      "children": Array [
        Object {
          "category": "graph",
          "isTransitive": true,
          "label": "h",
        },
        Object {
          "category": "graph",
          "isTransitive": true,
          "label": "h",
        },
      ],
      "isTransitive": true,
      "label": "(h ... h)",
    },
  ],
  "isTransitive": true,
  "label": "(g ... h)",
}
`;

exports[`parsing correctly produces the temporal and topological trees 2`] = `
Object {
  "category": "root",
  "children": Array [
    Object {
      "category": "sequence",
      "children": Array [
        Object {
          "category": "sequence",
          "children": Array [],
          "clique": "a",
          "label": "(g ... g): a",
        },
        Object {
          "category": "sequence",
          "children": Array [],
          "clique": "a",
          "label": "(h ... h): a",
        },
      ],
      "clique": "a",
      "label": "(g ... h): a",
    },
    Object {
      "category": "sequence",
      "children": Array [
        Object {
          "category": "sequence",
          "children": Array [],
          "clique": "b",
          "label": "(g ... g): b",
        },
        Object {
          "category": "sequence",
          "children": Array [],
          "clique": "c",
          "label": "(g ... g): c",
        },
        Object {
          "category": "sequence",
          "children": Array [],
          "clique": "c",
          "label": "(h ... h): c",
        },
      ],
      "clique": "c",
      "label": "(g ... h): c",
    },
  ],
  "label": "root",
}
`;
